Add API test cases and implement mock service

rtweera opened the pull request:
## Purpose
This PR adds test cases for API endpoints along with a mock service to simulate responses during testing. The goal is to ensure the API behaves as expected under various conditions, improving overall reliability and test coverage.

## Examples

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility

rtweera committed:
Add API endpoint test cases

Ensure all endpoints are behaving as expected and documented in the API
documentation.

rtweera committed:
Add mock service implementation for client tests

Add mock service for quick testing of the client.bal. This mock service
removes the server side dependency for the test cases and ensures they
run in a consistant manner.

rtweera committed:
Update and fix test cases for client testing.

Update positive and negative test cases for the client to handle certain
edge cases. This includes tests for invalid app id and payload with
complete settings (i.e. including all optionals).

Fix the test cases to deal with the DB delay happening at HubSpot's end.
HubSpot requires a delay around a minute to update the settings in the
DB. Test cases with method GET fails due to this delay, if it is not
accounted for.

rtweera committed:
Update and fix mock service for edge cases

Add appId validation to the mock service.
Fix correct responses for invalid appId.

rtweera committed:
Add updated Dependencies.toml

rtweera committed:
Merge pull request #4 from rtweera/test

Update tests with edge cases

Nuvindu commented on ballerina/tests/mock_service.bal:
```diff
@@ -0,0 +1,58 @@
+// AUTO-GENERATED FILE. DO NOT MODIFY.
+// This file is auto-generated by the Ballerina OpenAPI tool.
+
+// Copyright (c) 2025, WSO2 LLC. (http://www.wso2.com).
```
```suggestion
// Copyright (c) 2025, WSO2 LLC. (http://www.wso2.com).
```

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,139 @@
+    dependsOn: [testPutSettings]
+}
+function testGetSettings() returns error? {
+    if isLiveServer {
+        // Wait for the server to be updated the settings
+        runtime:sleep(60);
+    }
```
Do we need this?

rtweera replied:
If we are testing the live server, unfortunately we have to use a delay because it takes about a minute for their (hubspot's) databases to update the changes made by a `put` or a `delete`.

rtweera replied:
For an example if we do a `delete` and call a `get` right after that, we get the deleted content instead of the expected `no content` response. 

Same goes for a `put` followed by a `get`. We get a `204 No content` even though we added some through `put`.

This results in tests involving `get` to fail.

rtweera replied:
This results in tests involving `get` to fail.

ThisaruGuruge replied:
Are we sure it will be reflected within a minute? Otherwise this could lead to intermittent failures, which beats the purpose of the test. Is there a way to know when the DB is updated, without blindly waiting for a minute?

rtweera replied:
From their API, it seems there is no way to check. 

I will however go through this again today and get back to you, if I find any other way.

rtweera replied:
I have removed the delay. This will result in couple of tests failing (involving `get`) when testing against the live server, due to the said delay on the HubSpot's end.

All tests passes when tested against the mock service.

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,139 @@
+    }
+    ExternalSettings|http:Response settings = check hubSpotVideoConferrencing->/[appIdSigned32]();
+    test:assertTrue(settings is ExternalSettings, "Type mismatch");
+    if (settings is ExternalSettings) {
```
```suggestion
    if settings is ExternalSettings {
```

Nuvindu replied:
check for other places

rtweera replied:
Sure

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,139 @@
+    test:assertTrue(response.statusCode == 204, "Error deleting settings");
+}
+
+
+
```
```suggestion
```

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,139 @@
+import ballerina/http;
```
Let's add the license header here

Nuvindu replied:
And use the `Format Document` option to format the file

rtweera replied:
Added & formatted the entire module

rtweera committed:
Update the license header in mock_service.bal file

Remove the auto-generated comment from mock_service.bal file.

rtweera committed:
Fix extra parentheses in test.bal

Remove additional parentheses in if conditions in test.bal

rtweera committed:
Fix spelling mistakes in test.bal

rtweera committed:
Remove debug print statement from test.bal

Print statement used for checking the type of a response was removed
as it is not needed anymore.

rtweera committed:
Add license header to test.bal

rtweera committed:
Bal format the entire module

rtweera committed:
Add missing license for client, types, utils files

rtweera committed:
Add updated Dependencies.toml file

rtweera committed:
Merge pull request #5 from rtweera/test

Update with formatting fixes

ThisaruGuruge commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,151 @@
+}
+
+@test:Config {
+    enable: true
```
We don't need this. Tests are enabled by default. Check similar places.

rtweera replied:
Fixed.

rtweera committed:
Remove unnecessary test configurations

Tests are enabled by default, hence remove the unnecessary enable
configuration.

rtweera committed:
Merge pull request #6 from rtweera/test

Remove unnecessary test configurations

rtweera committed:
Remove delay in live tests for get method

Adding a delay in the tests could cause intermittent failures if the
delay is not sufficient for the particular situation. This commit
removes the delay in the live tests for the get method.

This will result in couple of test failures in the live tests but consistent
test results are achieved.

rtweera committed:
Add test case descriptions

Add concise test case descriptions with stating the positive or negative
scenario.

rtweera committed:
Remove unused import ballerina/lang.runtime

Previously used for delaying the live server test cases for get requests.
Since the delay was removed, this import is no longer needed.

rtweera committed:
Remove unused import ballerina/lang.runtime

Previously used for delaying the live server test cases for get requests.
Since the delay was removed, this import is no longer needed.

rtweera committed:
Merge branch origin test into test

rtweera committed:
Merge pull request #7 from rtweera/test

Remove delay in tests & Add test descriptions

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,141 @@
+    dependsOn: [testPutSettings]
+}
+function testGetIncorrectAppId() returns error? {
+    ExternalSettings|http:ClientRequestError|error settings = hubSpotVideoConferencing->/[1234]();
```
Shall we use variables instead of directly hardcoding values?

Nuvindu replied:
Check for other places

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,141 @@
+}
+function testDeleteSettingsAgain() returns error? {
+    http:Response response = check hubSpotVideoConferencing->/[appIdSigned32].delete();
+    test:assertTrue(response.statusCode == 204, "Error deleting settings");
```
Let's use `assertEquals()` here

rtweera committed:
Fix improper use of assertTrue in test.bal

Use assertEquals instead of assertTrue to compare the actual and
expected values when checking for the response code.

rtweera committed:
Use final variables for incorrectAppId in test.bal

Instead of hard-coded incorrectAppId in the request URL, use a final
variable for better readability and maintainability.

rtweera committed:
Merge pull request #8 from rtweera/test

Fix hardcoded value and incorrect use of assertTrue

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,142 @@
+
+configurable string liveServerUrl = "https://api.hubapi.com/crm/v3/extensions/videoconferencing/settings";
+configurable string localServerUrl = "http://localhost:9090";
+configurable boolean isLiveServer = false;
```
Let's make this variable as configurable or obtain the value from an env variable

rtweera replied:
Shall I move all three variable values (i.e. `liveServerUrl`, `localServerUrl`, `isLiveServer`) to `Config.toml` or just the `isLiveServer` one?

Nuvindu replied:
You can move all three

rtweera replied:
Noted and Thanks

rtweera committed:
Move all configurable values to Config.toml

With this change, configurable variable values can change without needing
a recompilation of the source code.

rtweera committed:
Merge pull request #9 from rtweera/test

Move all configurable values to Config.toml

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,142 @@
+final int:Signed32 appIdSigned32 = <int:Signed32>appId;
+final int:Signed32 incorrectAppId = 1234;
+final string serviceUrl = isLiveServer ? liveServerUrl : localServerUrl;
+final Client hubSpotVideoConferencing = check initClient();
```
We can just use `hubspot` as the client variable name

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,142 @@
+import ballerina/http;
+import ballerina/test;
+
+configurable string hapikey = ?;
+configurable int appId = ?;
+
+configurable string liveServerUrl = ?;
+configurable string localServerUrl = ?;
+configurable boolean isLiveServer = ?;
```
Let's add default values for these configurable variables

rtweera replied:
for `liveServerUrl` I added the actual url, instead of a dummy one. If it needs to be replaced to a dummy one, kindly let me know.

Nuvindu replied:
Can you add those default values in the file itself?

rtweera replied:
As in like this?

``` ballerina
configurable string hapikey = "my-key-123";
configurable int appId = 12345;

configurable string liveServerUrl = "https://api.hubapi.com/crm/v3/extensions/videoconferencing/settings";
configurable string localServerUrl = "http://localhost:9090";
configurable boolean isLiveServer = true;
```

Nuvindu replied:
Yeah. As discussed in the meeting

rtweera replied:
Yes sure

rtweera committed:
Add defaults for configurable variables in test.bal

rtweera committed:
Rename client to a simpler name

Instead of lengthy name `hubSpotVideoConferencing` for the client, use
the simpler name `hubspot`.

rtweera committed:
Merge pull request #10 from rtweera/test

Add configurable defaults and rename client

Nuvindu commented on ballerina/tests/test.bal:
```diff
@@ -0,0 +1,142 @@
+
+import ballerina/http;
+import ballerina/test;
+
+configurable string hapikey = "my-key-123";
+configurable int appId = 12345;
+
+configurable string liveServerUrl = "https://api.hubapi.com/crm/v3/extensions/videoconferencing/settings";
+configurable string localServerUrl = "http://localhost:9090";
+configurable boolean isLiveServer = true;
```
```suggestion
import ballerina/os;

configurable string hapikey = "my-key-123";
configurable int appId = 12345;

configurable string liveServerUrl = "https://api.hubapi.com/crm/v3/extensions/videoconferencing/settings";
configurable string localServerUrl = "http://localhost:9090";
configurable boolean isLiveServer = os:getEnv("IS_LIVE_SERVER") == "true";
```

Nuvindu committed:
Update variable to check for a live server
